modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/SearchResponse.scala

```diff
@@ -3,39 +3,67 @@ package ch.epfl.bluebrain.nexus.kg.service.routes
 import akka.http.scaladsl.model.{StatusCodes, Uri}
 import akka.http.scaladsl.server.Directives.{complete, extract, onSuccess}
 import akka.http.scaladsl.server.Route
+import cats.syntax.functor._
 import ch.epfl.bluebrain.nexus.commons.sparql.client.SparqlCirceSupport._
 import ch.epfl.bluebrain.nexus.kg.indexing.pagination.Pagination
 import ch.epfl.bluebrain.nexus.kg.indexing.query.QueryResult.{ScoredQueryResult, UnscoredQueryResult}
-import ch.epfl.bluebrain.nexus.kg.indexing.query.QueryResults
+import ch.epfl.bluebrain.nexus.kg.indexing.query.{QueryResult, QueryResults}
 import ch.epfl.bluebrain.nexus.kg.service.io.PrinterSettings._
 import ch.epfl.bluebrain.nexus.kg.service.query.LinksQueryResults
 import io.circe.Encoder
 import io.circe.generic.auto._
 
-import scala.concurrent.Future
+import scala.concurrent.{ExecutionContext, Future}
 
 trait SearchResponse {
 
   /**
     * Syntactic sugar for constructing a [[Route]] from the [[QueryResults]]
     */
-  implicit class QueryResultsOpts[A](qr: Future[QueryResults[A]]) {
+  implicit class QueryResultsOpts[Id](qr: Future[QueryResults[Id]]) {
+
+    private[routes] def addPagination(base: Uri, pagination: Pagination)(implicit
+                                                                         R: Encoder[UnscoredQueryResult[Id]],
+                                                                         S: Encoder[ScoredQueryResult[Id]]): Route =
+      extract(_.request.uri) { uri =>
+        onSuccess(qr) { result =>
+          val lqu = base.copy(path = uri.path, fragment = uri.fragment, rawQueryString = uri.rawQueryString)
+          complete(StatusCodes.OK -> LinksQueryResults(result, pagination, lqu))
+        }
+      }
 
     /**
       * Interface syntax to expose new functionality into [[QueryResults]] response type.
+      * Decides if either a generic type ''Id'' or ''Entity'' should be used.
       *
+      * @param fields     the fields query parameters
       * @param base       the service public uri + prefix
       * @param pagination the pagination values
       */
-    def buildResponse(base: Uri, pagination: Pagination)(implicit
-                                                         R: Encoder[UnscoredQueryResult[A]],
-                                                         S: Encoder[ScoredQueryResult[A]]): Route =
-      extract(_.request.uri) { uri =>
-        onSuccess(qr) { result =>
-          val lqu = base.copy(path = uri.path, fragment = uri.fragment, rawQueryString = uri.rawQueryString)
-          complete(StatusCodes.OK -> LinksQueryResults(result, pagination, lqu))
-        }
+    def buildResponse[Entity](fields: Set[String], base: Uri, pagination: Pagination)(
+        implicit
+        f: Id => Future[Option[Entity]],
```

Comment:
does this function need to be implicit?

Reply:
I think it is better, since it is a kind of mapping. We could create those implicits conversions directly in the `Main` however, or in the routes like it is right now.

But it is not strictly needed